caas/kubernetes/provider/proxy/setup.go

```diff
@@ -77,48 +68,136 @@ func CreateControllerProxy(
 			},
 		},
 	}
+)
+
+// proxyRoleForName builds the role needed for proxying to pods within a given
+// namespace.
+func proxyRoleForName(name string, lbs labels.Set) *rbacv1.Role {
+	role := proxyRole
+	role.ObjectMeta = meta.ObjectMeta{
+		Labels: lbs,
+		Name:   name,
+	}
+	return &role
+}
+
+// EnsureModelProxy ensures there is a proxy service account in existence for
+// the namespace of a Kubernetes model.
+func EnsureProxyService(
+	ctx context.Context,
+	lbs labels.Set,
+	name string,
+	roleI rbac.RoleInterface,
+	roleBindingI rbac.RoleBindingInterface,
+	saI core.ServiceAccountInterface,
+) error {
+	lbs = labels.Merge(lbs, utils.LabelsJuju)
+	pr := proxyRoleForName(name, lbs)
+	roleRVal, err := roleI.Create(ctx, pr, meta.CreateOptions{})
 
-	role, err := roleI.Create(context.TODO(), role, meta.CreateOptions{})
+	if k8serrors.IsAlreadyExists(err) {
+		roleRVal, err = roleI.Update(ctx, pr, meta.UpdateOptions{})
+	}
 	if err != nil {
-		return fmt.Errorf("creating proxy service account role: %w", err)
+		return errors.Annotate(err, "creating proxy service account role")
```

Comment:
`cannot create proxy service account role` would be better, same for other errors here